[1.21.4] Many API Additions

hammy275 opened the pull request:
A 1.20.4 port of #184 . This also adds a couple of new changes:

- Moved `isVanillaRenderPass()` to the bottom of the file, since its still marked `@Beta`. Perhaps we want to make a `VivecraftRenderingAPI` or something, just for organization?
- Added `isFirstRenderPass()` and `getCurrentRenderPass()`, as requested.

All features can be tested with [this test mod](https://github.com/hammy275/mc-vr-playground/releases/tag/v2.0.1). Note that the test mod itself has only been tested on Fabric, but _should_ work in Forge.

hammy275 committed:
Initial API

Fix Rebase Double-Tracker-Register Bug

Left-Handed API Rename

Rename API Impl Classes to Match API Classes

VRPose and VRBodyPart Rename

Separate Vivecraft Rendering API Interface

VR Capitalization Consistency

Migrate History to VRPoseHistory

FBT Additions (untested)

Move averageSpeed() to Impl Class

Hand Renaming and Docstring Changes

Vivecraft to VR Rename

Translate Position for Player in Common API

Not VR Active Conditional for Default Return

Expose getControllerRenderPos and setupRenderingAtController in API

Just Expose Hands

Docstring Improvements

Some Minor Adjustments

More Adjustments

Remove Unused Method

More Fixing Up

Adjust Names and Javadocs for Room Pose Client Methods

Fix Undoing Refactor Changes

VRPoseHistory Fixes

Merge VRBodyPart to One Enum

getInstance() --> instance() and Javadocs for a Couple

Refactor itemInUse() out to ItemInUseTracker

Make Not Default

Cache VRPoses and Only Make VRPoseHistory When Needed

Request Max Ticks of History Needed + Several VRPoseHistory Fixes

setTicksOfHistory --> requestTicksOfHistory

One Set Method Instead of Two

Javadoc Update

hammy275 commented:
Rebased onto `Multiloader-1.20.4`. Haven't tested it, but there were only a handful of merge conflicts with very simple fixes for all of them.

hammy275 commented on common/src/main/java/org/vivecraft/api/data/VRData.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    boolean isSeated();
+
+    /**
+     * @return Whether the player is using reversed hands.
+     */
+    boolean usingReversedHands();
+
```
Change this to be `isLeftHanded()` and update the javadoc accordingly.

hammy275 commented:
This PR is now caught up with the big refactor, but is untested. That said, I've marked it as draft, as there's several changes I'd like to make, and because the catching up to the refactor is entirely untested beyond ensuring it builds. These changes include:

- [ ] The renaming of reversed hands to left-handedness
- [ ] Refactoring to not have two versions of `VRData` and `VRDataImpl`. Ideally, just the interface and the class should exist, without needing to do internal conversions for the API.
- [ ] Rename `ClientAPIImpl` and `APIImpl` to `VivecraftClientAPIImpl` and `VivecraftAPIImpl`.
- [ ] Move rendering stuff out of the client API and into a "rendering API" area.
- [ ] Move haptic pulse triggering into `VRPlayer`, since it makes sense that once you have a `VRPlayer`, you should be able to trigger a haptic pulse on it.
- [ ] FBT support once merged.

hammy275 commented:
Squashed everything into one commit for now to make rebasing less annoying and now rebased on top of the JOML changes. Still need to test and make the changes above, but that should happen soon.

hammy275 commented:
For those following this, the migration to `VRPoseHistory` introduced the very stupid `deviceNum`. This is just a placeholder before I add FBT support and this gets replaced with a proper enum.

hammy275 commented:
Marking this as ready for review. The test mod is updated at https://github.com/hammy275/mc-vr-playground/releases/tag/v2.1.0, which should allow testing the vast majority of the API's functionality. I'm looking for all forms of feedback imaginable, whether it's bugs, feedback on API-design (big or small), or anything else.

There is definitely more I want to see added to the API, such as what's mentioned in https://github.com/Vivecraft/VivecraftMod/issues/335, but this PR is already really big in-scope, so I'd rather get this merged so future, smaller API additions can be made separately than continue to delay this PR.

fayer3 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,76 @@
+    TrackerTickType tickType();
+
+    /**
+     * Called for the client player if this tracker is currently causing the item to be used to not release the use key.
```
this is written a bit confusingly.
maybe better  to write it somethjigng like this
> Called for the client player, to check if this tracker ...

since it is called regardless of any tracker  state

hammy275 replied:
Should be resolved in https://github.com/Vivecraft/VivecraftMod/pull/255/commits/c3460107d1be491e58febef545201a8fa099648b.

fayer3 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    /**
+     * Called to reset data for the local player. This is called whenever {@link #isActive(LocalPlayer)} returns false.
```
maybe specify that this is meant to reset the tracker state, not reset some player data

fayer3 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    /**
+     * Called for all players, whether the tracker is active or not for them. This runs before
```
makes little sense to have a `all players` here, it still only runs for the local player

fayer3 commented on common/src/main/java/org/vivecraft/api/client/VRClientAPI.java:
```diff
@@ -0,0 +1,156 @@
+    void registerTracker(Tracker tracker);
+
+    /**
+     * Gets the VR pose representing the player in the room before the game tick.
```
hm, this data isn't actually bound by tick, it's regenerated every frame, and has the latest room data from the vr runtime
wheras the world pre/post ones are actually updated on tick

hammy275 replied:
Resolved! Will break the MC VR Playground test mod until I update that, but won't be doing that until all changes from here are addressed and tested.

fayer3 commented on common/src/main/java/org/vivecraft/api/client/VRClientAPI.java:
```diff
@@ -0,0 +1,156 @@
+
+    /**
+     * Gets the VR pose representing the player in Minecraft world coordinates before the game tick.
+     * This is the same as {@link #getPreTickRoomPose()} with translation to in-game world coordinates as of the last
```
not true, see the previous message. this is updated before tick, whereas the room one is not

hammy275 replied:
Unlike `vrdata_room_pre`, from a "Find Usages" on `vrdata_room_post`, this seems to only get updated at the end of the tick in `VRPlayer#postTick()`.

hammy275 replied:
Will likely adjust this name and the name of the method above, then. Likely something like `getLatestRoomPose()`/`getLatestWorldPose()` or similar, and leave the post-tick ones as-is.

fayer3 replied:
for world it is actually updated on tick, just the room one is not

fayer3 commented on common/src/main/java/org/vivecraft/api/client/VRClientAPI.java:
```diff
@@ -0,0 +1,156 @@
+    VRPose getPostTickWorldPose();
+
+    /**
+     * Gets the VR pose representing the player in Minecraft world coordinates after the game tick interpolated for
```
maybe missing a comma after `game tick`, or just strip the `after the game tick` entierly

fayer3 commented on common/src/main/java/org/vivecraft/api/client/VRRenderingAPI.java:
```diff
@@ -0,0 +1,52 @@
+    RenderPass getCurrentRenderPass();
+
+    /**
+     * @return Whether the current render pass is the first one it performed for this render cycle.
```
I think that `it` shouldn't be there

fayer3 commented on common/src/main/java/org/vivecraft/api/client/VRRenderingAPI.java:
```diff
@@ -0,0 +1,52 @@
+    Vec3 getHandRenderPos(InteractionHand hand);
+
+    /**
+     * Sets the provided {@link PoseStack} to render at the position of the provided {@link InteractionHand}.
```
this also sets the rotation to be aligned with the controller

fayer3 commented on common/src/main/java/org/vivecraft/api/data/VRBodyPart.java:
maybe this should be combined with the one used for networking, if done, the order needs to be the same though, so HMD would need to be last, to not break stuff

hammy275 replied:
Agreed. Will check to see if that would cause any other breakages. If not, I'll rename the networking one to `VRBodyPart` and expose that.

hammy275 replied:
Looks to be good and pushed! I'm thinking we may want to refactor a bit so HMD can be at the top, perhaps adding a method to get the `VRBodyPart` by `InteractionHand` and a method for getting it by a 0 or 1 index.
Or alternatively, having HMD be third in the list rather than the last.

hammy275 replied:
Going to not, since it shouldn't matter much at the end of the day where it is in the enum.

fayer3 commented on common/src/main/java/org/vivecraft/api/client/data/VRPoseHistory.java:
```diff
@@ -0,0 +1,97 @@
+     * Passing a value larger than this number to any methods below in their maxTicksBack or ticksBack parameters
+     * will throw an {@link IllegalArgumentException}.
+     */
+    int MAX_TICKS_BACK = 20;
```
is 20 Ticks (1 second) not too short?

hammy275 replied:
I can't imagine people wanting to use longer history, since the main intention is less-so for viewing literal historical data and moreso for using it to predict future movements, determine speed, etc.

That said, it might make sense to include an ability to specify how much history you need as a mod, and Vivecraft will just keep history for the largest number a mod asks for.

fayer3 replied:
just as a reference, the history classes we have internally have 450 frame samples, which is 5 seconds at 90hz

hammy275 replied:
Sounds good. Will set it to 5 seconds worth of ticks.

fayer3 commented on common/src/main/java/org/vivecraft/client/api_impl/data/VRPoseHistoryImpl.java:
```diff
@@ -0,0 +1,146 @@
+    }
+
+    @Override
+    public double averageSpeed(VRBodyPart bodyPart, int maxTicksBack) throws IllegalArgumentException {
```
this isn't right, you need to check all samples here,
making a circle that returns to the origin again would return 0 speed elsewise

hammy275 replied:
Definitely, yeah. I think my intention with initially writing this method was to only do the current tick and the provided tick, but it really should be all ticks from the current back up to `maxTicksBack`, not just current and that many ticks ago.

hammy275 replied:
Wait, are you referring to `averageVelocity()` or `averageSpeed()` with this?

fayer3 replied:
speed, the velocity should cancel itself out anyway

hammy275 replied:
Think I'm going to rewrite all of the methods in this file to use the history going up to `maxTicksBack`, rather than just the current and the provided tick value (where applicable). Should create more accurate data.

hammy275 replied:
Pushed a commit that should fix this. It's untested though, so may have issues that I haven't caught. Will resolve those with another round of bug fixing once this review is fully addressed.

fayer3 commented on common/src/main/java/org/vivecraft/client/api_impl/data/VRPoseHistoryImpl.java:
```diff
@@ -0,0 +1,146 @@
+        if (currentData == null) {
+            return null;
+        }
+        ListIterator<VRPose> iterator = this.dataQueue.listIterator(this.dataQueue.size() - 1);
```
why do the iteration starting from the back?
also, shouldn't this start at `iters` if done like that?
like it is now, it will always do the full time duration, not till `maxTicksBack`

hammy275 replied:
I think this code was written when the last item in the queue was the most recent, and I forgot to update it to be the other way around. Will fix.

hammy275 replied:
See https://github.com/Vivecraft/VivecraftMod/pull/255#discussion_r1919313489

fayer3 commented on common/src/main/java/org/vivecraft/client_vr/ClientDataHolderVR.java:
```diff
@@ -139,4 +144,12 @@ public void registerTracker(Tracker tracker) throws IllegalArgumentException {
+     * @return Whether some tracker is currently using an item.
+     */
+    public boolean isTrackerUsingItem(LocalPlayer player) {
+        return ClientDataHolderVR.getInstance().getTrackers().stream().anyMatch(tracker -> tracker.itemInUse(player));
```
why call `getInstance`, if this is called on the instance

hammy275 replied:
Not sure why I did that, is resolved now.

fayer3 commented on common/src/main/java/org/vivecraft/client_vr/VRData.java:
```diff
@@ -502,6 +534,16 @@ public Matrix4f getMatrix() {
+            return new VRBodyPartDataImpl(
+                getPosition(),
+                new Vec3(getDirection()),
+                quat
```
why is the quat creation not also just here?

hammy275 replied:
Good point. Fixed!

fayer3 commented on common/src/main/java/org/vivecraft/client_vr/gameplay/VRPlayer.java:
```diff
@@ -306,11 +301,11 @@ public void preRender(float partialTick) {
+        for (Tracker tracker : ClientDataHolderVR.getInstance().getTrackers())
+        {
+            if (tracker.tickType() == Tracker.TrackerTickType.PER_FRAME)
+            {
```
why are those brackets on a new line

hammy275 replied:
I think this is code that survived through some copy-pasting from before Vivecraft's codebase was formatted. This is fixed now, though.

fayer3 commented on common/src/main/java/org/vivecraft/client_vr/gameplay/VRPlayer.java:
```diff
@@ -405,13 +400,13 @@ public void tick(LocalPlayer player) {
+        {
+            if (tracker.tickType() == Tracker.TrackerTickType.PER_TICK)
+            {
+                tracker.idleTick(mc.player);
```
my are these using mc.player? or is that a remnace of before the refactor

hammy275 replied:
I can't recall what was changed surrounding `Tracker`s from after the refactor. What do you mean here (and further, I do see the passed `LocalPlayer` can be `null`, do `Tracker`s run outside of being in the world?)

fayer3 replied:
I mean why is it not using the player provided to the tick method, like it does in the refactor

hammy275 replied:
Likely me missing that when dealing with merge conflicts. Will fix that, thank you for catching it.

fayer3 commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/RowTracker.java:
```diff
@@ -21,24 +23,29 @@ public class RowTracker extends Tracker {
-            return false;
-        } else if (!(player.getVehicle() instanceof Boat)) {
-            return false;
+        } else if (p != null && p.isAlive()) {
```
why is that changed? is that also a pre refactor thing?

hammy275 replied:
Almost certainly. Likely messed up when dealing with merge conflicts 😅

fayer3 commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/SwingTracker.java:
```diff
@@ -390,14 +396,19 @@ else if (blockstate.getBlock() instanceof NoteBlock ||
         // TODO set destroyTicks to 1 to cancel multiple sound events per hit
-        // MCReflection.PlayerController_blockHitDelay.set(Minecraft.getInstance().gameMode, 0);
-        // Minecraft.getInstance().gameMode.blockBreakingCooldown = 1;
+        // MCReflection.PlayerController_blockHitDelay.set(this.mc.gameMode, 0);
```
same refactor thing probably

hammy275 replied:
That one was intentional, since the use of `Minecraft.getInstance()` was what was historically there before the tracker had access to `this.mc`. It does look like I missed updating a comment for that, though.

fayer3 commented on common/src/main/java/org/vivecraft/common/api_impl/data/VRBodyPartDataImpl.java:
```diff
@@ -0,0 +1,54 @@
+import org.joml.Quaternionfc;
+import org.vivecraft.api.data.VRBodyPartData;
+
+public class VRBodyPartDataImpl implements VRBodyPartData {
```
maybe make this a record?

hammy275 replied:
This is a good point. Already did it with `VRPoseImpl`, should do it here as well. 'Tis done now.

fayer3 commented on common/src/main/java/org/vivecraft/common/api_impl/data/VRBodyPartDataImpl.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Override
+    public String toString() {
+        return "Position: " + getPos() + "\tRotation: " + getRot();
```
why use a tab character here? I would just separate them with a comma or newlines

hammy275 replied:
Good point, this is a comma now.

fayer3 commented on common/src/main/java/org/vivecraft/common/network/FBTMode.java:
any reason these were deleted, and recreated, instead of moved?

hammy275 replied:
I think git decided too much of the file had changed. When I was doing it, I did literally move the file.

fayer3 commented on common/src/main/java/org/vivecraft/common/network/Pose.java:
```diff
@@ -32,4 +37,17 @@ public void serialize(FriendlyByteBuf buffer) {
+        return new VRBodyPartDataImpl(fromVector3fc(this.position).add(playerPos), fromVector3fc(this.orientation.transform(MathUtils.BACK, new Vector3f())), this.orientation);
+    }
+
+    private static Vec3 fromVector3fc(Vector3fc vec) {
```
since that is implemented in multiple classes, it should be moved to the MathHelper

fayer3 commented on common/src/main/java/org/vivecraft/client/ClientVRPlayers.java:
```diff
@@ -531,5 +535,37 @@ public float getBodyYawRad() {
+            : null;
+    }
+
+    private static Vec3 fromVector3fc(Vector3fc vec) {
```
since that is implemented in multiple classes, it should be moved to the MathHelper

fayer3 commented on common/src/main/java/org/vivecraft/common/api_impl/data/VRPoseImpl.java:
any reason this isn't just using VrPlayerState?

hammy275 replied:
This is a good point. Would make a lot of sense to have that just implement the interface, then let the other two make instances of that.

hammy275 replied:
On further investigation, since `VrPlayerState` stores positions relative to the player position, while the API gives world positions, this is unideal. I talk more about it [in the server](https://discord.com/channels/923042654537805874/923188682909958185/1329848133710250025) but the tl;dr is either:
- `VrPlayerState` ends up storing both relative positions and world positions, which seems bad, because it's then impossible to tell which is being worked with at a glance.
- `VrPlayerState` ends up storing the player position to make the API happy, which causes a lot of issues surrounding talking with legacy Vivecraft versions.

Curious to hear if you think I should do some refactoring to achieve one of the two above or not. I personally lean against it, but I'm happy to do it if you think it'd be better.

fayer3 replied:
didn't see that, if that is the case, then they should be separate

fayer3 replied:
they just looked the same on first inspection

hammy275 replied:
Totally get that, looked the same to me on first inspection too.

fayer3 commented on common/src/main/java/org/vivecraft/server/ServerVivePlayer.java:
```diff
@@ -162,4 +180,21 @@ public void setVR(boolean vr) {
+        return this.vrPlayerState.leftHanded();
+    }
+
+    public VRPose asVRPose(Vec3 playerPos) {
```
this could just use the ServerPlayer position of this object

fayer3 commented on common/src/main/java/org/vivecraft/server/ServerVivePlayer.java:
```diff
@@ -142,6 +147,19 @@ public Vec3 getBodyPartPos(BodyPart bodyPart) {
+     * @param c controller to get the rotation for
+     * @return controller rotation as a quaternion
+     */
+    public Quaternionfc getControllerQuaternion(int c) {
```
what is this for? seems unused

hammy275 replied:
Likely stuck around from before the FBT rework. Removed as it goes unused, as you mentioned.

fayer3 reviewed (commented):
looked though, and added some notes

fayer3 commented on common/src/main/java/org/vivecraft/common/utils/MathUtils.java:
```diff
@@ -218,4 +218,13 @@ public static float bodyYawRad(Vector3fc rightHand, Vector3fc leftHand, Vector3f
     }
+
+    /**
+     * Converst a {@link Vector3fc} to a {@link Vec3}.
```
typo here, should be converts

hammy275 replied:
This is a devastating blow to my programming career /j

In all seriousness, thank you for catching that.

hammy275 commented:
The review as it was should be addressed at this point. Going to update MC VR Playground so some testing can be done on the PR as it is.

hammy275 commented:
Small thing: going to rename all the `getInstance()` calls to `instance()` so it's a bit less word-y while still communicating the same thing.

hammy275 commented:
The above has been done, and [a new version of the test mod has been made](https://github.com/hammy275/mc-vr-playground/releases/tag/v2.2.0) to go with. Beyond updating to fit the newest state of the PR, it also mostly tests the other functions in `VRPoseHistory`.

This PR should be ready for another round of review if wanted, or to go forward if no further review is deemed necessary.

hammy275 commented:
Based on talk in Discord:

- [ ] Move `itemInUse()` out of `Tracker` to a separate `ItemInUseTracker` interface. This interface is checked for during `Tracker` registration to be additionally held in a separate list for `itemInUse` checks.
- [ ] Add a `Tracker` that uses `itemInUse()` to the test mod.

hammy275 commented:
The above is done! Updated version at https://github.com/hammy275/mc-vr-playground/releases/tag/v2.3.0

hammy275 commented:
From some more discussion:

- [ ] `VRPoseHistory` is populated even when no one is using the API. We should only start populating it when some mod calls the method to get the history for the first time.
- [ ] Conversions to the API's data (`VRPose` and its `VRBodyPartData`s) should be cached, so we don't create multiple `VRPose` instances of the same data if multiple mods want access to the same data.

hammy275 commented:
Changed it so the longest pose history needed by a mod is the pose history kept. If no pose history is requested by any mod, no pose history is kept. [Test mod has been updated accordingly](https://github.com/hammy275/mc-vr-playground/releases/tag/v2.3.1), so this PR is back in a "ready to review/merge/whatever else" state,

hammy275 commented:
[New test mod release](https://github.com/hammy275/mc-vr-playground/releases/tag/v2.3.2) for the rename above.

fayer3 committed:
some formatting, renaming, javadoc

hammy275 committed:
Off-by-one and averagePosition Fixes in VRPoseHistory

hammy275 committed:
VRPoseHistory on Server and Available for Other Clients on Client

hammy275 committed:
Javadoc Updating and getHistoricalData return null when history not available

hammy275 committed:
Clearer Error When No Pose History Requested

hammy275 committed:
Fake Player Checking in API

hammy275 committed:
Historical VR Poses VRAPI Note

hammy275 committed:
Remove Unused Import

fayer3 committed:
interact modules

fayer3 committed:
rename InteractModule methods, reword some javadoc

fayer3 committed:
add some hands

fayer3 committed:
remove redundant javadoc

fayer3 committed:
mention haptic on isActive

fayer3 committed:
add server side haptic trigger

fayer3 committed:
adjust haptic javadoc

fayer3 committed:
add since to all api javadoc
restore the legacy api with a deprecation warning and make it link to the new one
remove redundant VRPose.getHand

fayer3 committed:
remove beta qualifier, and clarify that the input for setupRenderingAtHand is expected to be an identity

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/ClientDataHolderVR.java:
```diff
@@ -165,6 +173,26 @@ public void registerTracker(Tracker tracker) throws IllegalArgumentException {
+                "InteractModule '" + module.getId() + "' is already added and should not be added again!");
+        }
+        this.interactModules.add(module);
+        this.interactTracker.setModules(this.interactModules);
```
This means that for each module added this will use the full list of modules, not just the latest added one. Is this fine?

fayer3 replied:
yeah I added that, this updates the api modules in the interact tracker, which makes a new list of pre api, api and post api modules

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/interact_modules/BlockInteractionModule.java:
```diff
@@ -0,0 +1,131 @@
+            ))
+            {
+                return true;
+            } else if (player.getItemInHand(hand).getItem() == Items.BUCKET && blockstate.liquid()) {
```
Maybe not for now, but potential use the loader api for modded fluid containers. Will be more work to abstract but could work.

ferriarnus commented on common/src/main/java/org/vivecraft/client_vr/gameplay/trackers/InteractTracker.java:
```diff
@@ -2,66 +2,65 @@
+        this.entityModule = new EntityInteractionModule(mc, dh);
+        this.blockModule = new BlockInteractionModule(mc, dh);
+
+        this.preAPIModules = List.of(this.hotbarModule, this.bowModule, this.thirdCamModule, this.screenCamModule);
```
I wouldn't have pre and post modules. Make all use the api and use the priority to sort them? While technically all modules should be in between, there might be a niche use case to not do so?

fayer3 replied:
I discussed that with hammy, the pre api ones should definetly always run first imo, the post api ones could use the default priority

ferriarnus commented on common/src/main/java/org/vivecraft/api/VRAPI.java:
```diff
@@ -0,0 +1,126 @@
+     * @throws IllegalArgumentException If a non-positive number is supplied.
+     * @since 1.3.0
+     */
+    void requestTicksOfHistory(int maxTicksBack) throws IllegalArgumentException;
```
I've not paid enough attention to this pr, but why us this separate from `getHistoricalVRPoses`? As in `getHistoricalVRPoses(Player player, int maxTicksBack)`. Is there a reason to use them not following each other?

fayer3 replied:
this is meant to be called once, to indicate to us how much data we need to keep around

ferriarnus reviewed (commented):
Had a quick look, thank you for keeping this going! Just some small questions/comments.

hammy275 committed:
Remove requestTicksOfHistory(), just keep 200 ticks

fayer3 committed:
fix interactmodule.onRelease being called on every reset

fayer3 committed:
move bowtracker keypress logic to the interact module

fayer3 committed:
move interact modules instances to dataholder
move register logic to interact tracker
remove postAPIModules and just add them to the api ones

fayer3 committed:
rename the bow interaction bow module to BowModule, to be in line with the tracker name

fayer3 committed:
move interact modules to regular api list
change API tracker/interact registers to take a varargs
clarify that negative values are allowed for interact priorities and what range Vivecraft uses

fayer3 committed:
Merge remote-tracking branch 'remotes/origin/Multiloader-1.21.4' into vivecraft-api-1.20.4

fayer3 committed:
move tracker/interactmodule registration to a separate event

fayer3 committed:
move api event to after initial reload, to be guaranteed after the mod initializers

fayer3 committed:
make ItemInUseTracker extend Tracker

fayer3 committed:
Merge remote-tracking branch 'remotes/origin/Multiloader-1.21.4' into vivecraft-api-1.20.4

fayer3 committed:
add debug rendering for some trackers

hammy275 commented on common/src/main/java/org/vivecraft/api/client/ItemInUseTracker.java:
```diff
@@ -0,0 +1,23 @@
+
+/**
+ * An interface that should be implemented by {@link Tracker}s if they want to take advantage of
+ * {@link #itemInUse(LocalPlayer)}.
```
Should update this comment to clarify that it's a type of `Tracker`, not something to be implemented alongside `Tracker`.

hammy275 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,72 @@
+
+/**
+ * A tracker is an object that is run for the local player during the game tick or before rendering a frame only if
+ * they are in VR. Using trackers are one of the cleanest ways to interact with Vivecraft's data; it's how Vivecraft
```
Probably on me, but I think replacing the ";" with a "," here is the right call.
```suggestion
 * they are in VR. Using trackers are one of the cleanest ways to interact with Vivecraft's data, it's how Vivecraft
```

fayer3 replied:
also, should it be `Using trackers is one of the ...` or `Trackers are one of the ...`, the current gramar is not correct I think

fayer3 replied:
for now I changed it to `is`

fayer3 replied:
can reopen this if you think it should be the other one

hammy275 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * Whether the tracker is active for the local player.
+     *
+     * @param player Player being checked if they are active for this tracker instances. Will be {@code null} when not in a world.
```
Typo
```suggestion
     * @param player Player being checked if they are active for this tracker instance. Will be {@code null} when not in a world.
```

hammy275 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,72 @@
+    void doProcess(@Nullable LocalPlayer player);
+
+    /**
+     * The ticking type for this tracker.<br>
```
```suggestion
     * The ticking type for this tracker.
     * <br>
```

hammy275 commented on common/src/main/java/org/vivecraft/api/client/Tracker.java:
```diff
@@ -0,0 +1,72 @@
+
+    /**
+     * The ticking type for this tracker.<br>
+     * If this is {@link TrackerTickType#PER_FRAME}, the tracker is called once with the local player per frame before the frame is rendered.<br>
```
I think the `<br>` on this line makes sense to stay where it is, since for those viewing the unformatted Javadoc, this shouldn't have a gap in the middle. Your call if you'd like to move it down for consistency, though.

hammy275 commented on common/src/main/java/org/vivecraft/api/client/VRClientAPI.java:
```diff
@@ -0,0 +1,183 @@
+    /**
+     * Gets the VR pose representing the player in the room after the most recent poll of VR hardware.
+     *
+     * @return The most up-to-date VR pose representing the player in the room, or null if the local player isn't in VR.
```
Won't leave a suggestion for all of these, but should replace the instances of `null` that are on their own with `{@code null}`

hammy275 commented on common/src/main/java/org/vivecraft/api/client/VivecraftRegistrationEvent.java:
```diff
@@ -0,0 +1,22 @@
+     * @param module InteractModules to register.
+     * @since 1.3.0
+     */
+    void registerInteractModule(InteractModule... module);
```
Nitpick: Rename `tracker` and `module` to `trackers` and `modules`? Just a nitpick, feel free to ignore.

fayer3 replied:
you mean the methofd name? or the parameter

hammy275 commented on common/src/main/java/org/vivecraft/api/data/FBTMode.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    ARMS_ONLY,
+    /**
+     * head, hands, waist, and feet trackers are available.
```
```suggestion
     * Head, hands, waist, and feet trackers are available.
```

hammy275 commented on common/src/main/java/org/vivecraft/api/data/FBTMode.java:
```diff
@@ -0,0 +1,32 @@
+     */
+    ARMS_LEGS,
+    /**
+     * head, hands, waist, feet, elbow, and knee trackers are available.
```
```suggestion
     * Head, hands, waist, feet, elbow, and knee trackers are available.
```

hammy275 reviewed (commented):
Mainly looked at the Javadocs for any last-minute typo fixes

fayer3 committed:
javadoc updates

fayer3 committed:
make register method plural

hammy275 committed:
Move RenderPass to API, add Descriptions for Many Methods, and Adjust Some

fayer3 committed:
add since 1.20.5 to the setupRenderingAtHand with a matrix

hammy275 committed:
amount of ticks to number of ticks

hammy275 committed:
Many Tracker Changes

fayer3 committed:
some javavdoc rewording.
rename Tracker.reset to Tracker.inactiveProcess

fayer3 committed:
remove duplicate ProcessType definitions

fayer3 committed:
formatting
add some missing method descriptions

fayer3 committed:
move processType()

fayer3 committed:
got to was

hammy275 committed:
idleProcess javadoc update

hammy275 committed:
or to and

fayer3 committed:
use VRPose there, instead of pose/position

fayer3 committed:
rename registration event to client registration event

fayer3 committed:
move fake player check to ServerVRPlayers

hammy275 committed:
Keyboard API Improvements

fayer3 committed:
move keyboard enums to data package, use keyboard enums in native vivecraft code

fayer3 committed:
move isVRActive/isVRInitialized top the top in the client api

fayer3 committed:
move historical poses after regular poses, and make the impl reflect the same order

fayer3 committed:
move server haptics to dedicated server api

fayer3 committed:
rename server side haptics to sendHapticPulse
